pkg/streamingpromql/functions/common.go

```diff
@@ -43,3 +46,7 @@ func FloatTransformationDropHistogramsFunc(transform func(f float64) float64) In
 func Passthrough(seriesData types.InstantVectorSeriesData, _ *pooling.LimitingPool) (types.InstantVectorSeriesData, error) {
 	return seriesData, nil
 }
+
+// RangeVectorStepFunction is a function to operate on a range vector step.
+// floatBuffer and histogramBuffer will contain the values for the step range.
+type RangeVectorStepFunction func(step types.RangeVectorStepData, rangeSeconds float64, floatBuffer *types.FPointRingBuffer, histogramBuffer *types.HPointRingBuffer) (*promql.FPoint, *promql.HPoint, error)
```

Comment:
One more thought: what if we split this in two, one for computing the output over float samples and one for the same over native histograms? Or have a simple case for most functions perhaps?

With the exception of `count_over_time` (and maybe `last_over_time`?), I don't believe there are any functions over range vectors that support both floats and histograms at once for a single output point, so this should simplify the implementation of most of these functions somewhat.

Reply:
(Also `present_over_time`?)

Reply:
I think they may need to consider what happens when both a histogram and float exists a single point. For example `rate` will return nil points where both exist. I don't think we should split these up at this point, I think it is more logical for the operating function to consider how to handle when both are present rather than move that logic up into the definition of the function.

Reply:
This is how I'm thinking about it: there three categories of functions over range vectors:

* functions that only work with floats and ignore native histograms, even if they occur in a single step where floats are also present: `min_over_time`, `max_over_time`, `quantile_over_time`, `stddev_over_time`, `stdvar_over_time`, `mad_over_time`, `changes`, `deriv`, `holt_winters`, `idelta`, `irate`, `predict_linear`

* functions that handle a single step seeing both floats and native histograms: `count_over_time`, `resets`, `last_over_time`, `present_over_time`

* functions that will not emit a sample if the step has both floats and native histograms, and otherwise produce a sample matching the input type (eg. float input produces float output and histogram input produces histogram output): `delta`, `increase`, `rate`, `avg_over_time`, `sum_over_time`

I'm wondering if we should have specific cases for each of these three that handle the common functionality shared by each group, so that the function-specific implementations can be very simple.

Reply:
> I'm wondering if we should have specific cases for each of these three that handle the common functionality shared by each group, so that the function-specific implementations can be very simple.

I guess my concern here is how do we determine the cases? We'd still need `RangeVectorStepFunc` in `FunctionOverRangeVector` or otherwise have two separate funcs like `RangeVectorFloatStepFunc` and `RangeVectorHistogramStepFunc`, perhaps with a enum for each of the cases you describe.

However we'd likely still want the function to get the actual points as otherwise we're handling too many different states in `FunctionOverRangeVector` (imo).

> * functions that only work with floats and ignore native histograms, even if they occur in a single step where floats are also present: `min_over_time`, `max_over_time`, `quantile_over_time`, `stddev_over_time`, `stdvar_over_time`, `mad_over_time`, `changes`, `deriv`, `holt_winters`, `idelta`, `irate`, `predict_linear`

Wouldn't these just call `floatBuffer.UnsafePoints(step.RangeEnd)` and never bother with the histograms returning nil?

> * functions that handle a single step seeing both floats and native histograms: `count_over_time`, `resets`, `last_over_time`, `present_over_time`

These need both, so any re-arrangements are moot for this case.

> * functions that will not emit a sample if the step has both floats and native histograms, and otherwise produce a sample matching the input type (eg. float input produces float output and histogram input produces histogram output): `delta`, `increase`, `rate`, `avg_over_time`, `sum_over_time`

Unless `FunctionOverRangeVector` checks for the presence of both floats+histograms, the functions would still need both buffers to check themselves. Having FunctionOverRangeVector check doesn't make as much sense to me.

I think either I don't fully understand what you're envisioning here or I disagree with it. Would it be okay if we leave this for now and revisit it in a followup PR if necessary?

Reply:
I think this will be easier to discuss once we implement some more of these functions - let's pause this discussion until we get to that.